vnext/Microsoft.ReactNative/CoalescingEventEmitter.h

```diff
@@ -0,0 +1,49 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "CoalescingEventEmitter.g.h"
+#include "JSValue.h"
+#include "ReactHost/React.h"
+#include "ReactPropertyBag.h"
+#include "winrt/Microsoft.ReactNative.h"
+
+#include <deque>
+#include <mutex>
+
+namespace winrt::Microsoft::ReactNative::implementation {
+
+struct CoalescingEvent
+{
+  std::optional<int64_t> coalesceKey;
+  bool expired{false};
+  folly::dynamic args;
+};
+
+struct CoalescingEventEmitter : CoalescingEventEmitterT<CoalescingEventEmitter> {
+ public:
+  static ReactPropertyId<CoalescingEventEmitter> Property() noexcept;
+  static winrt::Microsoft::ReactNative::CoalescingEventEmitter FromContext(IReactContext const & context) noexcept;
+  static winrt::Microsoft::ReactNative::CoalescingEventEmitter FromContext(Mso::React::IReactContext const & context) noexcept;
+
+  CoalescingEventEmitter(Mso::CntPtr<Mso::React::IReactContext> &&context) noexcept;
+
+  void EmitJSEvent(int64_t coalesceKey, JSValueArgWriter const & paramsArgWriter) noexcept;
```

Comment:
I'm not in love with how this allows collisions. Need to think a bit more about the coalesce-key API.

One bit I was playing with was also partitioning by event name as part of it, since we don't likely want to coalesce between different events.